test/hotspot/jtreg/compiler/lib/template_framework/DataName.java

```diff
@@ -0,0 +1,224 @@
+                if (supertype != null && !supertype.isSubtypeOf(dn.type())) { return false; }
+                return true;
+            };
+        }
+
+        /**
+         * Create a filtered {@link View}, where all {@link DataName}s must be subtypes of {@code type}.
+         *
+         * @param type The type of which all {@link DataName}s must be subtypes of.
+         * @return The filtered {@link View}.
+         * @throws UnsupportedOperationException If this {@link View} was already filtered with
+         *                                       {@link subtypeOf} or {@link exactOf}.
+         */
+        public View subtypeOf(DataName.Type type) {
+            if (subtype != null) {
+                throw new UnsupportedOperationException("Cannot constrain to subtype " + type + ", is already constrained: " + subtype);
+            }
+            return new View(mutability, type, supertype);
+        }
+
+        /**
+         * Create a filtered {@link View}, where all {@link DataName}s must be subtypes of {@code type}.
+         *
+         * @param type The type of which all {@link DataName}s must be supertype of.
+         * @return The filtered {@link View}.
+         * @throws UnsupportedOperationException If this {@link View} was already filtered with
+         *                                       {@link supertypeOf} or {@link exactOf}.
+         */
+        public View supertypeOf(DataName.Type type) {
+            if (supertype != null) {
+                throw new UnsupportedOperationException("Cannot constrain to supertype " + type + ", is already constrained: " + supertype);
+            }
+            return new View(mutability, subtype, type);
+        }
+
+        /**
+         * Create a filtered {@link View}, where all {@link DataName}s must be of exact {@code type},
+         * hence it must be both subtype and supertype thereof.
+         *
+         * @param type The type of which all {@link DataName}s must be.
+         * @return The filtered {@link View}.
+         * @throws UnsupportedOperationException If this {@link View} was already filtered with
+         *                                       {@link subtypeOf}, {@link supertypeOf} or {@link exactOf}.
+         */
+        public View exactOf(DataName.Type type) {
+            return subtypeOf(type).supertypeOf(type);
+        }
+
+        /**
+         * Samples a random {@link DataName} from the filtered set, according to the weights
+         * of the contained {@link DataName}s.
+         *
+         * @return The sampled {@link DataName}.
+         * @throws UnsupportedOperationException If the type was not constrained with either of
+         *                                       {@link subtypeOf}, {@link supertypeOf} or {@link exactOf}.
+         * @throws RendererException If the set was empty.
+         */
+        public DataName sample() {
+            DataName n = (DataName)Renderer.getCurrent().sampleName(predicate());
```

Comment:
Do you really need this cast? Can't you just return a `Name`. From the uses it seems that you only call interface methods from `Name` at the use-sites.

Reply:
That would require `Name` to become public. I wanted to avoid that.

I want the user to think that `DataName` and `StructuralName` are separate but parallel. But internally, they have a unified implementation with `Name`.
An alternative would have been to use `NameSet` generically, once with `DataName` and `StructuralName`. But that would mean we could have a `DataName` with the same `name` as a `StructuralName`, because they would be in separate `NameSet`s.

What do you think?